File: 03-04-2025_Rotten_Oranges.java
public int orangesRotting(int[][] mat) {
    int m = mat.length;
    int n = mat[0].length;
    
    Queue<int[]> q = new LinkedList<>();
    int count = 0, count2 = 0;

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (mat[i][j] == 2) {
                q.add(new int[]{i, j});
            } else if (mat[i][j] == 1) {
                count2++;
            }
        }
    }

    int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    while (!q.isEmpty()) {
        int flag = 0;
        int size = q.size();

        for (int z = 0; z < size; z++) {
            int[] temp = q.poll();
            int i = temp[0], j = temp[1];

            for (int[] dir : directions) {
                int ni = i + dir[0], nj = j + dir[1];

                if (ni >= 0 && nj >= 0 && ni < m && nj < n && mat[ni][nj] == 1) {
                    mat[ni][nj] = 2;
                    q.add(new int[]{ni, nj});
                    flag = 1;
                    count2--;
                }
            }
        }

        if (flag == 1) {
            count++;
        }
    }

    return count2 == 0 ? count : -1;
}

File: 03-04-2025_Rotten_Oranges.cpp
bool isValid(int i, int j, int m, int n){
    if(i < 0 || j < 0 || i >= m || j >= n)
        return false;
    return true;
}

int orangesRotting(vector<vector<int>>& arr) {
    // Code here
    int m = arr.size();
    int n = arr[0].size();
    
    queue <pair<int, int>> q;
    // vector <vector<int>> visited(m, vector<int>(n, 0));
    int count = 0;
    int count2 = 0;

    for(int i = 0; i < m; i++){
        for(int j = 0; j < n; j++){
            if(arr[i][j] == 2){
                q.push({i,j});
                // visited[i][j] = 1;
            }

            else if(arr[i][j] == 1)
                count2++;
        }
    }



    while(!q.empty()){
        // count++;
        int flag = 0;
        int z = q.size();

        while(z--){
            pair <int, int> temp = q.front();
            q.pop();
            bool a, b, c, d;
            int i = temp.first;
            int j = temp.second;
            // visited[i][j] = 1;
            a = isValid(temp.first, temp.second+1, m, n);
            b = isValid(temp.first, temp.second-1, m, n);
            c = isValid(temp.first+1, temp.second, m, n);
            d = isValid(temp.first-1, temp.second, m, n);

            if(a && arr[i][j+1] == 1){
                arr[i][j+1] = 2;
                q.push({i, j+1});
                flag = 1;
                count2--;
            }
            if(b && arr[i][j-1] == 1){
                arr[i][j-1] = 2;
                q.push({i, j-1});
                flag = 1;
                count2--;
            }
            if(c && arr[i+1][j] == 1){
                arr[i+1][j] = 2;
                q.push({i+1, j});
                flag = 1;
                count2--;
            }
            if(d && arr[i-1][j] == 1){
                arr[i-1][j] = 2;
                q.push({i-1, j});
                flag = 1;
                count2--;
            }
        }

        if(flag)
            count++;
    }

    // for(int i = 0; i < m; i++){
    //     for(int j = 0; j < n; j++)
    //         if(arr[i][j] == 1)
    //             return -1;
    // }

    if(count2 != 0)
        return -1;

    return count;
}
